### PR TITLE
test: setup pre-commit + initial unit test suite

### PR DESCRIPTION
**Pull request by cbuto**
This PR adds some initial setup of unit tests and pre-commit (linting, etc).

The general idea behind the unit tests is to test things that should be easy to validate before applying any manifests to a running cluster (that feedback loop is long and time consuming). We can validate the general folder structure of the repo and also go as far as validating each files contents where it makes sense.

A few more ideas for tests:
1. Test for the helm repo folder
2. Validate the metadata.yaml structure (make sure each service's metadata file contains `displayName`, `description`, `category`, `scope`, `overview`, and `icon`)
3. Validate that each helm release has an associated helm repo defined
4. Validate that each helm release references a `ConfigMap` that exists in the `defaults` directory 

In the next PR, I'll be setting up mergebot/TC

### CODE REVIEW

**Comment by cwyl02** [comment]
Likely an issue on my laptop.. when I run `make test` I see it tries to install gotestsum but bash still complain about `gotestsum` doesn't exist.

```
{13:25}~/workspace/dkp-catalog-applications:cbuto/linter-test-setup ✗ ➭ make test                                                                                         :cbuto/linter-test-setup|…1⚑2
Executing target: install-tools
cd tools && go install -v github.com/golangci/golangci-lint/cmd/golangci-lint github.com/onsi/ginkgo/v2 gotest.tools/gotestsum
Executing target: test
cd tests && gotestsum \
                --junitfile junit-report.xml \
                --junitfile-testsuite-name=relative \
                --junitfile-testcase-classname=short \
                -- \
                -covermode=atomic \
                -coverprofile=coverage.out \
                -race \
                -short \
                -v \
                ./...
/bin/bash: line 1: gotestsum: command not found
```

**Review by cwyl02 — APPROVED** [review]

**Comment by cwyl02** [comment]
approving since `make test` seems to work fine on @gracedo 's laptop.

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	cd tests && go tool cover -html=coverage.out -o coverage.html
+
+.PHONY: mod-tidy
+mod-tidy: ## go mod tidy
+	$(call print-target)
+	cd tests && go mod tidy
+	cd tools && go mod tidy
+
+.PHONY: diff
+diff: ## git diff
```
Comment (gracedo) [inline Makefile:55]: just wondering if we need this one? and is the below `RES` line supposed to `exit 1` there? (isn't clear to me what that line is accomplishing)

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,56 @@
+repos:
+- repo: local
```
Comment (gracedo) [inline .pre-commit-config.yaml:2]: I liked the metadata.yaml validation that max added in k-apps: https://github.com/mesosphere/kommander-applications/pull/127/files#diff-63a9c44a44acf85fea213a857769990937107cf072831e1a26808cfde9d096b9R5-R11 can we add the same here?

**Review by gracedo — COMMENTED** [review]
Exciting stuff! Just a few comments, otherwise lgtm, I tested out most of the make targets
```
$ make
clean                          remove files created during build
diff                           git diff
go-clean                       go clean build, test and modules caches
install-tools                  go install tools
lint                           golangci-lint
mod-tidy                       go mod tidy
pre-commit                     Runs pre-commit on all files
test                           go test with race detector and code coverage
```

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+.PHONY: lint
+lint: ## golangci-lint
+lint: install-tools
+	$(call print-target)
+	cd tests && golangci-lint run -c ${REPO_ROOT}/.golangci.yml --fix
+
+.PHONY: test
+test: ## go test with race detector and code coverage
+test: install-tools
```
Comment (takirala) [inline Makefile:33]: need to install gotestsum as a dep. see https://github.com/mesosphere/kommander/blob/main/make/tools.mk#L71-L72

**Comment by cbuto** [comment]
> Maybe an issue on my laptop.. when I run `make test` I see it tries to install gotestsum but bash still complain about `gotestsum` doesn't exist.
> 
> ```
> {13:25}~/workspace/dkp-catalog-applications:cbuto/linter-test-setup ✗ ➭ make test                                                                                         :cbuto/linter-test-setup|…1⚑2
> Executing target: install-tools
> cd tools && go install -v github.com/golangci/golangci-lint/cmd/golangci-lint github.com/onsi/ginkgo/v2 gotest.tools/gotestsum
> Executing target: test
> cd tests && gotestsum \
>                 --junitfile junit-report.xml \
>                 --junitfile-testsuite-name=relative \
>                 --junitfile-testcase-classname=short \
>                 -- \
>                 -covermode=atomic \
>                 -coverprofile=coverage.out \
>                 -race \
>                 -short \
>                 -v \
>                 ./...
> /bin/bash: line 1: gotestsum: command not found
> ```

@cwyl02 I think I see what happened, should be fixed in 78b3b5d26f2668cf4bc84466fc2db814317ffbc3. can you try now?

File: Makefile (hunk 3 above)
Comment (cbuto) [reply Makefile:33]: I took the `install-tools` target from https://github.com/mesosphere/golang-repository-template, which ends up doing something similar but installs all the tools at once. I did have to fix it in this commit 78b3b5d26f2668cf4bc84466fc2db814317ffbc3 to fix the gotestsum issue i think

```
cd tools && go install -v $(shell cd tools && go list -f '{{ join .Imports " " }}' -tags=tools)
```

File: Makefile (hunk 1 above)
Comment (cbuto) [reply Makefile:55]: probably not! I started this Makefile from https://github.com/mesosphere/golang-repository-template which had this target, I'll take it out!

Comment (cbuto) [reply Makefile]: removed in 9467b06fb7032f7c92b30592d0c1a2b192c4b681

File: .pre-commit-config.yaml (hunk 2 above)
Comment (cbuto) [reply .pre-commit-config.yaml:2]: for sure, added in 9467b06fb7032f7c92b30592d0c1a2b192c4b681

**Comment by cwyl02** [comment]
@cbuto turns out to be missing $GOBIN in my $PATH. now `make test` works for me

**Review by gracedo — APPROVED** [review]

File: Makefile (hunk 3 above)
Comment (gracedo) [reply Makefile:33]: resolving, can revisit in follow-up pr if needed

**Pull request merged**